alf/pretrained_models/model_adapters.py

```diff
@@ -0,0 +1,252 @@
+            return
+        m = self._m[0]
+        m.weight.data.add_(-self._adapter_weight())
+        m.forward = m._unmerged_forward
+        self._merged = False
+
+    def detach(self):
+        """Detach from and recover the base module.
+
+        Note that this operation is *irreversible*. Once detached, there is no
+        way to add the adapter back to the module.
+
+        We still keep the adapter weights after detachment.
+        """
+        if self._merged:
+            self.unmerge()
+        m = self._m[0]
+        m.forward = m._forward0
+        del m._forward0
+        del m._unmerged_forward
+
+    @classmethod
+    def can_adapt(cls, m: nn.Module) -> bool:
+        """Check if the adapter class can adapt a given module.
+        """
+        raise NotImplementedError()
+
+    @classmethod
+    def _decompose_weight_dims(cls, m: nn.Module) -> Tuple[int]:
+        """Return two (reshaped) dimensions ``(nrows, ncols)`` so that
+        ``nrows*ncols == np.prod(m.weight.shape)``.
+
+        The reinterpreted shape will be used to create LoRA weights ``self._wA``
+        and ``self._wB``.
+
+        Returns:
+            tuple: a pair of ints ``(nrows, ncols)``.
+        """
+        raise NotImplementedError()
+
+    def forward(self, input):
+        raise NotImplementedError()
+
+
+@alf.configurable
+class LinearAdapter(LoRA):
+    """Adapter for linear layers.
+    """
+
+    @classmethod
+    def can_adapt(cls, m: nn.Module) -> bool:
+        return isinstance(m, nn.Linear)
+
+    @classmethod
+    def _decompose_weight_dims(cls, m: nn.Module):
+        return m.out_features, m.in_features
+
+    def forward(self, input):
+        return F.linear(input, self._adapter_weight())
+
+
+@alf.configurable
+class Conv2dAdapter(LoRA):
+    """Adapter for Conv2d layers.
+
+    The most natural way of a LoRA decomposition for Conv2d is
+
+    .. code-block:: python
+
+        (rank, kernel_size[0] * kernel_size[1] * in_channels) x (out_channels, rank)
+
+    However, since ``out_channels`` is usually small, this decomposition is not
+    low-rank actually and it won't save much memory.
+
+    We can first reinterpret the weight matrix as a shape of
+
+    .. code-block:: python
+
+        (out_channels * kernel_size[0], in_channels * kernel_size[1])
+
+    to make the in- and out-dimensions balanced, and then decompose.
+    """
+
+    @classmethod
+    def can_adapt(cls, m: nn.Module) -> bool:
+        return isinstance(m, nn.Conv2d)
+
+    @classmethod
+    def _decompose_weight_dims(cls, m: nn.Module):
+        kernel_size = m.kernel_size
+        if not isinstance(kernel_size, tuple):
+            kernel_size = (kernel_size, ) * 2
+        nrows = m.out_channels // m.groups * kernel_size[0]
+        ncols = m.in_channels * kernel_size[1]
+        return nrows, ncols
+
+    def forward(self, input):
+        m = self._m[0]
+        return F.conv2d(
+            input,
+            self._adapter_weight(),
```

Comment:
the third argument of F.conv2d is `bias`

Reply:
> the third argument of F.conv2d is `bias`

Right, but because it's not used so it's skipped here.